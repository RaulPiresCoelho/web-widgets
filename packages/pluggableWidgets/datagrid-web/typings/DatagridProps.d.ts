/**
 * This file was generated from Datagrid.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { ComponentType, CSSProperties, ReactNode } from "react";
<<<<<<< HEAD
import { ActionValue, DynamicValue, EditableValue, ListValue, ListActionValue, ListAttributeValue, ListAttributeListValue, ListExpressionValue, ListReferenceValue, ListReferenceSetValue, ListWidgetValue, SelectionSingleValue, SelectionMultiValue } from "mendix";
=======
import { ActionValue, DynamicValue, EditableValue, ListValue, ListActionValue, ListAttributeValue, ListExpressionValue, ListReferenceValue, ListReferenceSetValue, ListWidgetValue, SelectionSingleValue, SelectionMultiValue } from "mendix";
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
import { Big } from "big.js";

export type ItemSelectionMethodEnum = "checkbox" | "rowClick";

export type ItemSelectionModeEnum = "toggle" | "clear";

<<<<<<< HEAD
export type LoadingTypeEnum = "spinner" | "skeleton";

=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
export type ShowContentAsEnum = "attribute" | "dynamicText" | "customContent";

export type HidableEnum = "yes" | "hidden" | "no";

export type WidthEnum = "autoFill" | "autoFit" | "manual";

export type MinWidthEnum = "auto" | "minContent" | "manual";

export type AlignmentEnum = "left" | "center" | "right";

export interface ColumnsType {
    showContentAs: ShowContentAsEnum;
<<<<<<< HEAD
    attribute?: ListAttributeValue<string | Big | boolean | Date> | ListAttributeListValue<string | Big | boolean | Date>;
=======
    attribute?: ListAttributeValue<string | Big | boolean | Date>;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    content?: ListWidgetValue;
    dynamicText?: ListExpressionValue<string>;
    exportValue?: ListExpressionValue<string>;
    header?: DynamicValue<string>;
    tooltip?: ListExpressionValue<string>;
    filter?: ReactNode;
    filterAssociation?: ListReferenceValue | ListReferenceSetValue;
    filterAssociationOptions?: ListValue;
<<<<<<< HEAD
    fetchOptionsLazy: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    filterAssociationOptionLabel?: ListExpressionValue<string>;
    visible: DynamicValue<boolean>;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    allowEventPropagation: boolean;
    width: WidthEnum;
    minWidth: MinWidthEnum;
    minWidthLimit: number;
    size: number;
    alignment: AlignmentEnum;
    columnClass?: ListExpressionValue<string>;
    wrapText: boolean;
}

export type PaginationEnum = "buttons" | "virtualScrolling" | "loadMore";

<<<<<<< HEAD
export type ShowPagingButtonsEnum = "always" | "auto";

export type PagingPositionEnum = "bottom" | "top" | "both";

=======
export type PagingPositionEnum = "bottom" | "top" | "both";

export type ShowPagingButtonsEnum = "always" | "auto";

>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
export type ShowEmptyPlaceholderEnum = "none" | "custom";

export type OnClickTriggerEnum = "single" | "double";

export type ConfigurationStorageTypeEnum = "attribute" | "localStorage";

export interface FilterListType {
    filter: ListAttributeValue<string | Big | boolean | Date>;
}

export interface ColumnsPreviewType {
    showContentAs: ShowContentAsEnum;
    attribute: string;
    content: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    dynamicText: string;
    exportValue: string;
    header: string;
    tooltip: string;
    filter: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    filterAssociation: string;
    filterAssociationOptions: {} | { caption: string } | { type: string } | null;
<<<<<<< HEAD
    fetchOptionsLazy: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    filterAssociationOptionLabel: string;
    visible: string;
    sortable: boolean;
    resizable: boolean;
    draggable: boolean;
    hidable: HidableEnum;
    allowEventPropagation: boolean;
    width: WidthEnum;
    minWidth: MinWidthEnum;
    minWidthLimit: number | null;
    size: number | null;
    alignment: AlignmentEnum;
    columnClass: string;
    wrapText: boolean;
}

export interface FilterListPreviewType {
    filter: string;
}

export interface DatagridContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    advanced: boolean;
    datasource: ListValue;
    refreshInterval: number;
    itemSelection?: SelectionSingleValue | SelectionMultiValue;
    itemSelectionMethod: ItemSelectionMethodEnum;
    itemSelectionMode: ItemSelectionModeEnum;
    showSelectAllToggle: boolean;
<<<<<<< HEAD
    loadingType: LoadingTypeEnum;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    columns: ColumnsType[];
    columnsFilterable: boolean;
    pageSize: number;
    pagination: PaginationEnum;
<<<<<<< HEAD
    showPagingButtons: ShowPagingButtonsEnum;
    showNumberOfRows: boolean;
    pagingPosition: PagingPositionEnum;
=======
    pagingPosition: PagingPositionEnum;
    showPagingButtons: ShowPagingButtonsEnum;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    loadMoreButtonCaption?: DynamicValue<string>;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder?: ReactNode;
    rowClass?: ListExpressionValue<string>;
    onClickTrigger: OnClickTriggerEnum;
    onClick?: ListActionValue;
    onSelectionChange?: ActionValue;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationStorageType: ConfigurationStorageTypeEnum;
    configurationAttribute?: EditableValue<string>;
<<<<<<< HEAD
    storeFiltersInPersonalization: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    filterList: FilterListType[];
    filtersPlaceholder?: ReactNode;
    filterSectionTitle?: DynamicValue<string>;
    exportDialogLabel?: DynamicValue<string>;
    cancelExportLabel?: DynamicValue<string>;
    selectRowLabel?: DynamicValue<string>;
}

export interface DatagridPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
<<<<<<< HEAD
    renderMode?: "design" | "xray" | "structure";
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    advanced: boolean;
    datasource: {} | { caption: string } | { type: string } | null;
    refreshInterval: number | null;
    itemSelection: "None" | "Single" | "Multi";
    itemSelectionMethod: ItemSelectionMethodEnum;
    itemSelectionMode: ItemSelectionModeEnum;
    showSelectAllToggle: boolean;
<<<<<<< HEAD
    loadingType: LoadingTypeEnum;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    columns: ColumnsPreviewType[];
    columnsFilterable: boolean;
    pageSize: number | null;
    pagination: PaginationEnum;
<<<<<<< HEAD
    showPagingButtons: ShowPagingButtonsEnum;
    showNumberOfRows: boolean;
    pagingPosition: PagingPositionEnum;
=======
    pagingPosition: PagingPositionEnum;
    showPagingButtons: ShowPagingButtonsEnum;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    loadMoreButtonCaption: string;
    showEmptyPlaceholder: ShowEmptyPlaceholderEnum;
    emptyPlaceholder: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    rowClass: string;
    onClickTrigger: OnClickTriggerEnum;
    onClick: {} | null;
    onSelectionChange: {} | null;
    columnsSortable: boolean;
    columnsResizable: boolean;
    columnsDraggable: boolean;
    columnsHidable: boolean;
    configurationStorageType: ConfigurationStorageTypeEnum;
    configurationAttribute: string;
<<<<<<< HEAD
    storeFiltersInPersonalization: boolean;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    onConfigurationChange: {} | null;
    filterList: FilterListPreviewType[];
    filtersPlaceholder: { widgetCount: number; renderer: ComponentType<{ children: ReactNode; caption?: string }> };
    filterSectionTitle: string;
    exportDialogLabel: string;
    cancelExportLabel: string;
    selectRowLabel: string;
}
