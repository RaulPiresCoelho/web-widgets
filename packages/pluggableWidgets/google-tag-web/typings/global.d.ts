type DojoConnectHandle = string & {__dojo__: string};

declare global {
    interface Window {
        dataLayer: any[];
        mx: {
            ui: {
                getContentForm(): { path: string; url: string; title: string };
            };
            session: {
                getSessionObjectId(): string,
<<<<<<< HEAD
                getConfig() : { locale: { code: string } },
                getUserId(): string
=======
                getConfig() : { locale: { code: string } }
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            }
        };
        dojo: {
            connect: (o: object, f: string, cb: () => void) => DojoConnectHandle;
            disconnect: (DojoConnectHandle) => void;
        }
    }
    interface Document {
        mxGtag?: {
            ensureGtagIncluded: (tagId: string) => boolean;
            getGtag: () => (...args: any[]) => void;
        };
    }
}

export {};
