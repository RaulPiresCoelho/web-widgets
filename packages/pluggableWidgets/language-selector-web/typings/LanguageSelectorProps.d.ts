/**
 * This file was generated from LanguageSelector.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { CSSProperties } from "react";
<<<<<<< HEAD
import { DynamicValue, ListValue, ListExpressionValue } from "mendix";
=======
import { ListValue, ListExpressionValue } from "mendix";
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)

export type PositionEnum = "left" | "right" | "top" | "bottom";

export type TriggerEnum = "click" | "hover";

export interface LanguageSelectorContainerProps {
    name: string;
    class: string;
    style?: CSSProperties;
    tabIndex?: number;
    languageOptions: ListValue;
    languageCaption: ListExpressionValue<string>;
    position: PositionEnum;
    trigger: TriggerEnum;
    hideForSingle: boolean;
<<<<<<< HEAD
    screenReaderLabelCaption?: DynamicValue<string>;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}

export interface LanguageSelectorPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
<<<<<<< HEAD
    renderMode?: "design" | "xray" | "structure";
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    languageOptions: {} | { caption: string } | { type: string } | null;
    languageCaption: string;
    position: PositionEnum;
    trigger: TriggerEnum;
    hideForSingle: boolean;
<<<<<<< HEAD
    screenReaderLabelCaption: string;
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}
