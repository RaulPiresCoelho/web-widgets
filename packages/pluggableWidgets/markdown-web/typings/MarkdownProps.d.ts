/**
 * This file was generated from Markdown.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { EditableValue } from "mendix";

export interface MarkdownContainerProps {
    name: string;
    tabIndex?: number;
    id: string;
    stringAttribute: EditableValue<string>;
}

export interface MarkdownPreviewProps {
    readOnly: boolean;
<<<<<<< HEAD
    renderMode?: "design" | "xray" | "structure";
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    stringAttribute: string;
}
