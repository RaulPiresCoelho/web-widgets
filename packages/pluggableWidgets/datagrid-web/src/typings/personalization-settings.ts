<<<<<<< HEAD
import { FilterData } from "@mendix/widget-plugin-filtering/typings/settings";
=======
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
import { ColumnId } from "./GridColumn";
import { SortDirection, SortRule } from "./sorting";

export interface ColumnPersonalizationSettings {
    columnId: ColumnId;
    size: number | undefined;
    hidden: boolean;
    orderWeight: number;
    sortDir: SortDirection | undefined;
    sortWeight: number | undefined;
<<<<<<< HEAD
=======
    filterSettings: unknown;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}

interface ColumnPersonalizationStorageSettings {
    columnId: ColumnId;
    size: number | undefined;
    hidden: boolean;
<<<<<<< HEAD
}

export type ColumnFilterSettings = Array<[key: ColumnId, data: FilterData]>;

export type GroupFilterSettings = Array<[key: string, data: FilterData]>;

export interface GridPersonalizationStorageSettings {
    name: string;
    schemaVersion: 2;
    settingsHash: string;
    columns: ColumnPersonalizationStorageSettings[];
    groupFilters: GroupFilterSettings;
    columnFilters: ColumnFilterSettings;
=======
    filterSettings: unknown;
}

export interface GridPersonalizationStorageSettings {
    name: string;
    schemaVersion: number;
    settingsHash: string;

    columns: ColumnPersonalizationStorageSettings[];
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
    columnOrder: ColumnId[];
    sortOrder: SortRule[];
}
