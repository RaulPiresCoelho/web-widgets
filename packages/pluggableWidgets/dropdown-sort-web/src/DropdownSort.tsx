<<<<<<< HEAD
import { observer } from "mobx-react-lite";
import { createElement, ReactElement, useRef } from "react";
import { useSortControl } from "@mendix/widget-plugin-sorting/helpers/useSortControl";
import { SortingStoreInterface } from "@mendix/widget-plugin-sorting/typings";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { DropdownSortContainerProps } from "../typings/DropdownSortProps";
import { SortComponent } from "./components/SortComponent";
import { withSortStore } from "./hocs/withSortStore";

function Container(props: DropdownSortContainerProps & { sortStore: SortingStoreInterface }): ReactElement {
    const id = (useRef<string>().current ??= `DropdownSort${generateUUID()}`);
    const sortProps = useSortControl(
        { ...props, emptyOptionCaption: props.emptyOptionCaption?.value },
        props.sortStore
    );

    return (
        <SortComponent
            className={props.class}
            placeholder={props.emptyOptionCaption?.value}
            id={id}
=======
import { Context, createElement, ReactElement, useCallback, useContext, useMemo, useRef } from "react";
import { ListAttributeValue } from "mendix";
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { SortContextValue, SortDirection, SortInstruction, getSortDispatcher } from "@mendix/widget-plugin-sorting";
import { DropdownSortContainerProps } from "../typings/DropdownSortProps";
import { SortComponent, SortOption } from "./components/SortComponent";

export function DropdownSort(props: DropdownSortContainerProps): ReactElement {
    const id = useRef(`DropdownSort${generateUUID()}`);

    const SortContext = (getSortDispatcher() ?? {}) as Context<SortContextValue>;
    const alertMessage = (
        <Alert bootstrapStyle="danger">
            The Drop-down sort widget must be placed inside the header of the Gallery widget.
        </Alert>
    );
    const { sortDispatcher, attributes, initialSort } = useContext(SortContext) ?? {};

    let defaultValue;
    let defaultDirection;

    const options = useMemo(
        () => [
            {
                caption: props.emptyOptionCaption?.value ?? "",
                value: ""
            },
            ...(attributes?.map(attribute => ({ caption: attribute.caption, value: attribute.attribute.id })) ?? [])
        ],
        [attributes, props.emptyOptionCaption?.value]
    );

    const updateSort = useCallback(
        (value: SortOption, direction: SortDirection): void => {
            const filteredOption = attributes?.find(attr => attr.attribute.id === value.value);
            sortDispatcher({
                getSortCondition: () => getSortCondition(filteredOption?.attribute, direction, value)
            });
        },
        [attributes, sortDispatcher]
    );

    if (!sortDispatcher || !attributes) {
        return alertMessage;
    }

    if (initialSort && initialSort.length > 0) {
        // Keeping the code in order to implement a future multi-sorting
        const [sort] = initialSort;
        const [id, direction] = sort;
        if (id) {
            defaultValue = options.find(option => option.value === String(id));
        }
        if (direction) {
            defaultDirection = direction;
        }
    }

    return (
        <SortComponent
            key={defaultValue?.value ?? "sort-component"}
            className={props.class}
            defaultDirection={defaultDirection}
            defaultOption={defaultValue}
            emptyOptionCaption={props.emptyOptionCaption?.value}
            id={id.current}
            options={options}
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            screenReaderButtonCaption={props.screenReaderButtonCaption?.value}
            screenReaderInputCaption={props.screenReaderInputCaption?.value}
            styles={props.style}
            tabIndex={props.tabIndex}
<<<<<<< HEAD
            {...sortProps}
=======
            updateSort={updateSort}
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        />
    );
}

<<<<<<< HEAD
const Widget = withSortStore(observer(Container));

export function DropdownSort(props: DropdownSortContainerProps): ReactElement {
    return <Widget {...props} />;
=======
function getSortCondition(
    listAttribute: ListAttributeValue | undefined,
    direction: SortDirection,
    value: SortOption
): SortInstruction | undefined {
    if (!listAttribute || !listAttribute.sortable || !value) {
        return undefined;
    }

    const { id } = listAttribute;

    return [id, direction];
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}
