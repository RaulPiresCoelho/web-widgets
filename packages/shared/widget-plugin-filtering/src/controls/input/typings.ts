<<<<<<< HEAD
import { InputFilterInterface } from "../../typings/InputFilterInterface";
import { InputStore } from "../../stores/InputStore";
import { AllFunctions } from "../../typings/FilterFunctions";
=======
export type Optional<T> = T | undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)

export interface BaseProps {
    adjustable: boolean;
    className?: string;
    id?: string;
    name: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    styles?: React.CSSProperties;
    tabIndex?: number;
<<<<<<< HEAD
    type: "text" | "number";
    badge?: string;
}

export type FilterFnList<T> = Array<{ value: T; label: string }>;

export interface InputProps<Fn extends AllFunctions> {
    onFilterChange: (filterFn: Fn) => void;
    filterFn: Fn;
    filterFnList: FilterFnList<Fn>;
    inputStores: [InputStore, InputStore];
    disableInputs?: boolean;
    inputRef: React.RefAttributes<HTMLInputElement>["ref"];
    defaultValue?: string;
}

export interface InputComponentProps<Fn extends AllFunctions> extends BaseProps, InputProps<Fn> {}

export interface InputHookProps<Fn, V> {
    filters: FilterFnList<Fn>;
    defaultFilter: Fn;
    defaultValue: V | undefined;
    changeDelay?: number;
    filterStore: InputFilterInterface;
    disableInputs?: (filterFn: Fn) => boolean;
=======
}

export type FilterList<T> = Array<{ value: T; label: string }>;

export interface InputProps<TFilterEnum> {
    onFilterChange: (type: TFilterEnum, isFromUserInteraction: boolean) => void;
    defaultFilter: TFilterEnum;
    filters: FilterList<TFilterEnum>;

    onInputChange: React.ChangeEventHandler<HTMLInputElement>;
    inputType: "text" | "number";
    inputDisabled?: boolean;
    inputRef?: React.RefAttributes<HTMLInputElement>["ref"];
    inputValue: string;
}

export interface InputComponentProps<TFilterEnum> extends BaseProps, InputProps<TFilterEnum> {}

export interface ValueHelper<T> {
    /** Function to map input string to value. */
    fromString(arg: string): T;
    /** Function to map value to string. */
    toString(arg: T): string;
    /** Function to compare the previous value and the new value. */
    equals(a: T, b: T): boolean;
}

export interface InputHookProps<TValue, TFilterEnum> {
    defaultFilter: TFilterEnum;
    filters: FilterList<TFilterEnum>;
    value: Optional<TValue>;
    onChange: (value: Optional<TValue>, filter: TFilterEnum) => void;
    valueHelper: ValueHelper<TValue>;
    changeDelay?: number;
    parentChannelName: string | null;
    name: string;
    inputDisabled?: (filter: TFilterEnum) => boolean;
    inputType: "text" | "number";
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
}
