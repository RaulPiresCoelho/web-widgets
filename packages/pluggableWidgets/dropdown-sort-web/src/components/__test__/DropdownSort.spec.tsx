<<<<<<< HEAD
import { FilterAPIv2 } from "@mendix/widget-plugin-filtering/context";
import { HeaderFiltersStore, HeaderFiltersStoreProps } from "@mendix/widget-plugin-filtering/stores/HeaderFiltersStore";
import { SortAPI } from "@mendix/widget-plugin-sorting/context";
import { SortAPIProvider, SortListType } from "@mendix/widget-plugin-sorting/providers/SortAPIProvider";
import { ListAttributeId } from "@mendix/widget-plugin-sorting/typings";
import { ListAttributeValueBuilder, dynamicValue } from "@mendix/widget-plugin-test-utils";
import { fireEvent, render } from "@testing-library/react";
import { ListValue } from "mendix";
import { createContext, createElement } from "react";
import { DropdownSortContainerProps } from "../../../typings/DropdownSortProps";
import { DropdownSort } from "../../DropdownSort";

const commonProps: DropdownSortContainerProps = {
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test"
};

export interface StaticInfo {
    name: string;
    filtersChannelName: string;
}

const headerFilterStoreInfo: StaticInfo = {
    name: commonProps.name,
    filtersChannelName: ""
};

// CONTEXT
const props: HeaderFiltersStoreProps = {
    filterList: []
};
const headerFilterStore = new HeaderFiltersStore(props, headerFilterStoreInfo, null);
(window as any)["com.mendix.widgets.web.filterable.filterContext.v2"] = createContext<FilterAPIv2>(
    headerFilterStore.context
);

const sortList: SortListType[] = [
    {
        attribute: new ListAttributeValueBuilder().withId("attribute1").withType("String").withSortable(true).build(),
        caption: dynamicValue<string>("Option 1")
    },
    {
        attribute: new ListAttributeValueBuilder().withId("attribute2").withType("Decimal").withSortable(true).build(),
        caption: dynamicValue<string>("Option 2")
    }
];
const sortProvider = new SortAPIProvider({
    datasource: { sortOrder: [[sortList[0].attribute.id as ListAttributeId, "asc"]] } as ListValue,
    sortList
});
(window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext<SortAPI>(sortProvider.context);
// END CONTEXT

=======
import { Alert } from "@mendix/widget-plugin-component-kit/Alert";
import { SortInstruction, SortContextValue } from "@mendix/widget-plugin-sorting";
import { ListAttributeValueBuilder } from "@mendix/widget-plugin-test-utils";
import { createContext, createElement } from "react";
import { DropdownSort } from "../../DropdownSort";
import { render } from "@testing-library/react";
import { mount } from "enzyme";
import { SortComponent } from "../SortComponent";

const commonProps = {
    class: "filter-custom-class",
    tabIndex: 0,
    name: "filter-test",
    sortOrder: "asc" as const
};

>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
describe("Dropdown Sort", () => {
    describe("with single instance", () => {
        afterEach(() => {
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });

        describe("with correct context", () => {
<<<<<<< HEAD
            it("loads correct values from attributes", () => {
                const filter = render(<DropdownSort {...commonProps} />);
                fireEvent.click(filter.getByRole("textbox"));

                const items = filter.getAllByRole("menuitem");

                items.forEach((item, index) => {
                    if (index === 0) {
                        return;
                    }
                    expect(item.textContent).toBe(`Option ${index}`);
                });
=======
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext({
                    sortDispatcher: jest.fn(),
                    attributes: [
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("String")
                                .withSortable(true)
                                .build(),
                            caption: "Option 1"
                        },
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("Decimal")
                                .withSortable(true)
                                .build(),
                            caption: "Option 2"
                        }
                    ]
                } as SortContextValue);
            });

            it("loads correct values from attributes", () => {
                const filter = mount(<DropdownSort {...commonProps} />);

                expect(filter.find(SortComponent).prop("options")).toStrictEqual([
                    {
                        caption: "",
                        value: ""
                    },
                    {
                        caption: "Option 1",
                        value: "attribute1"
                    },
                    {
                        caption: "Option 2",
                        value: "attribute2"
                    }
                ]);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });

            it("renders correctly", () => {
                const { asFragment } = render(<DropdownSort {...commonProps} />);
<<<<<<< HEAD
                expect(asFragment()).toMatchSnapshot();
            });
        });

        describe("with view state", () => {
            it("loads correct default option", () => {
                const filter = render(<DropdownSort {...commonProps} />);
                fireEvent.click(filter.getByRole("textbox"));

                expect(filter.getByRole("textbox").getAttribute("value")).toStrictEqual("Option 1");
=======

                expect(asFragment()).toMatchSnapshot();
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
            });
        });

        describe("with view state", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext({
                    sortDispatcher: jest.fn(),
                    attributes: [
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute1")
                                .withType("String")
                                .withSortable(true)
                                .build(),
                            caption: "Option 1"
                        },
                        {
                            attribute: new ListAttributeValueBuilder()
                                .withId("attribute2")
                                .withType("Decimal")
                                .withSortable(true)
                                .build(),
                            caption: "Option 2"
                        }
                    ],
                    initialSort: [["attribute2", "asc"] as SortInstruction]
                } as SortContextValue);
            });

            it("loads correct default option", () => {
                const filter = mount(<DropdownSort {...commonProps} />);

                expect(filter.find(SortComponent).prop("defaultOption")).toStrictEqual({
                    caption: "Option 2",
                    value: "attribute2"
                });
            });

            afterAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });

        describe("with no context", () => {
            beforeAll(() => {
                (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
            });

            it("renders error message", () => {
<<<<<<< HEAD
                const filter = render(<DropdownSort {...commonProps} />);
                expect(filter.container.querySelector(".alert")?.textContent).toBe("Out of context");
=======
                const filter = mount(<DropdownSort {...commonProps} />);

                expect(filter.find(Alert).text()).toBe(
                    "The Drop-down sort widget must be placed inside the header of the Gallery widget."
                );
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            });
        });
    });

    describe("with multiple instances", () => {
        beforeAll(() => {
<<<<<<< HEAD
            (window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext<SortAPI>(
                sortProvider.context
            );
=======
            (window as any)["com.mendix.widgets.web.sortable.sortContext"] = createContext({
                sortDispatcher: jest.fn(),
                attributes: [
                    {
                        attribute: new ListAttributeValueBuilder()
                            .withId("attribute1")
                            .withType("String")
                            .withSortable(true)
                            .build(),
                        caption: "Option 1"
                    },
                    {
                        attribute: new ListAttributeValueBuilder()
                            .withId("attribute2")
                            .withType("Decimal")
                            .withSortable(true)
                            .build(),
                        caption: "Option 2"
                    }
                ]
            } as SortContextValue);
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
        });

        it("renders with a unique id", () => {
            const { asFragment: fragment1 } = render(<DropdownSort {...commonProps} />);
            const { asFragment: fragment2 } = render(<DropdownSort {...commonProps} />);

<<<<<<< HEAD
=======
            console.warn(fragment1().querySelector("input")?.getAttribute("aria-controls"));

>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
            expect(fragment1().querySelector("input")?.getAttribute("aria-controls")).not.toBe(
                fragment2().querySelector("input")?.getAttribute("aria-controls")
            );
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.sortable.sortContext"] = undefined;
            delete (global as any)["com.mendix.widgets.web.UUID"];
        });
    });
});
