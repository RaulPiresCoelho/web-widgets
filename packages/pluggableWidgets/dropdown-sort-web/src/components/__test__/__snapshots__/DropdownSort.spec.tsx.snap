// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Sort with single instance with correct context renders correctly 1`] = `
<DocumentFragment>
  <div
    class="dropdown-container filter-custom-class"
    data-focusindex="0"
  >
    <div
      class="dropdown-triggerer-wrapper"
    >
      <input
        aria-controls="DropdownSort1-dropdown-list"
        aria-expanded="false"
        aria-haspopup="true"
        class="form-control dropdown-triggerer"
<<<<<<< HEAD
        readonly=""
        value="Option 1"
=======
        value=""
>>>>>>> daa3fce04 (Add DE localization to rich-text-web)
      />
      <button
        class="btn btn-default btn-sort icon-asc"
      />
    </div>
  </div>
</DocumentFragment>
`;
